base/multidimensional.jl

```diff
@@ -690,14 +687,15 @@ julia> diff(vec(a))
  12
 ```
 """
-function diff(A::AbstractMatrix; dims::Integer)
-    if dims == 1
-        [A[i+1,j] - A[i,j] for i=1:size(A,1)-1, j=1:size(A,2)]
-    elseif dims == 2
-        [A[i,j+1] - A[i,j] for i=1:size(A,1), j=1:size(A,2)-1]
-    else
-        throw(ArgumentError("dimension must be 1 or 2, got $dims"))
-    end
+function diff(a::AbstractArray; dims::Integer)
+  @assert !has_offset_axes(a)
+  1 <= dims <= ndims(a) || throw(ArgumentError("dimension out of range"))
+
+  r = axes(a)
+  r0 = ntuple(i -> i == dims ? Base.OneTo(last(r[i]) - 1) : r[i], ndims(a))
```

Comment:
No need for `Base.` qualifications inside of Base.

Reply:
And `OneTo` won't be correct for all `AbstractArrays`, so better use a more general range starting at `first(r[i])`. Also, below `UnitRange` could be built using the standard `a:b` syntax, starting at `r[i][2]`.

Reply:
Currently the code ensures that the input does not have offset axes, so `OneTo` should technically work, though it should be fairly straightforward to tweak things to allow offset axes.